lib/Dialect/RTL/RTLFolds.cpp

```diff
@@ -732,3 +732,219 @@ void WireOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
+  case ICmpPredicate::sge:
+    return lhs.sge(rhs);
+  case ICmpPredicate::ult:
+    return lhs.ult(rhs);
+  case ICmpPredicate::ule:
+    return lhs.ule(rhs);
+  case ICmpPredicate::ugt:
+    return lhs.ugt(rhs);
+  case ICmpPredicate::uge:
+    return lhs.uge(rhs);
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+// Returns the result of applying the predicate when the LHS and RHS are the
+// exact same value.
+static bool applyCmpPredicateToEqualOperands(ICmpPredicate predicate) {
+  switch (predicate) {
+  case ICmpPredicate::eq:
+  case ICmpPredicate::sle:
+  case ICmpPredicate::sge:
+  case ICmpPredicate::ule:
+  case ICmpPredicate::uge:
+    return true;
+  case ICmpPredicate::ne:
+  case ICmpPredicate::slt:
+  case ICmpPredicate::sgt:
+  case ICmpPredicate::ult:
+  case ICmpPredicate::ugt:
+    return false;
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
+
+  // gt a, a -> false
+  // gte a, a -> true
+  if (lhs() == rhs()) {
+    auto val = applyCmpPredicateToEqualOperands(predicate());
+    return IntegerAttr::get(getType(), val);
+  }
+
+  auto lhs = constants[0].dyn_cast_or_null<IntegerAttr>();
+  auto rhs = constants[1].dyn_cast_or_null<IntegerAttr>();
+
+  // gt 1, 2 -> false
+  if (lhs && rhs) {
+    auto val = applyCmpPredicate(predicate(), lhs.getValue(), rhs.getValue());
+    return IntegerAttr::get(getType(), val);
+  }
+
+  return {};
+}
+
+namespace {
+// Canonicalizes a ICmp with a single constant
+struct ICmpCanonicalizeConstant final : public OpRewritePattern<ICmpOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(ICmpOp op,
+                                PatternRewriter &rewriter) const override {
+
+    APInt lhs, rhs;
+
+    // icmp 1, x -> icmp x, 1
+    if (matchPattern(op.lhs(), m_RConstant(lhs))) {
+      assert(!matchPattern(op.rhs(), m_RConstant(rhs)) && "Should be folded");
+      rewriter.replaceOpWithNewOp<ICmpOp>(
+          op, ICmpOp::getFlippedPredicate(op.predicate()), op.rhs(), op.lhs());
+      return success();
+    }
+
+    // Canonicalize with RHS constant
+    if (matchPattern(op.rhs(), m_RConstant(rhs))) {
+      ConstantOp constant;
+
+      auto getConstant = [&rewriter, &op](APInt &&constant) -> Value {
+        return rewriter.create<ConstantOp>(op.getLoc(), constant);
+      };
+
+      auto replaceWith = [&rewriter, &op](ICmpPredicate predicate, Value lhs,
+                                          Value rhs) -> LogicalResult {
+        rewriter.replaceOpWithNewOp<ICmpOp>(op, predicate, lhs, rhs);
+        return success();
+      };
+
+      auto replaceWithConstant = [&rewriter,
+                                  &op](bool constant) -> LogicalResult {
+        rewriter.replaceOpWithNewOp<ConstantOp>(op, APInt(1, constant));
+        return success();
+      };
+
+      switch (op.predicate()) {
+      case ICmpPredicate::slt:
+        // x < max -> x != max
```

Comment:
Let me know if you think these comments are overkill

Reply:
Not overkill - they are great.